src/util/fields.py

```diff
@@ -511,49 +510,69 @@ def rename_field(cr, model, old, new, update_references=True, domain_adapter=Non
             "The field %r from the model %r is now a standard Odoo field, but it already existed in the database "
             "(coming from a non-standard module) and thus has been renamed to %r" % (new, model, custom_name)
         )
         add_to_migration_reports(msg, "Non-standard fields")
 
+    cr.execute("UPDATE ir_model_fields SET name=%s WHERE model=%s AND name=%s RETURNING id", (new, model, old))
+    [fid] = cr.fetchone() or [None]
+
     if fid:
+        # recreate the xmlids
         name = IMD_FIELD_PATTERN % (model.replace(".", "_"), new)
-        # In some cases the same field may be present on ir_model_data with both the double __ and single _ name
-        # version. To avoid conflicts (module, name) on the UPDATE below we keep only the double __ version
+
         cr.execute(
             """
              DELETE FROM ir_model_data
-                   WHERE id IN (SELECT unnest((array_agg(id ORDER BY id))[2:count(id)])
-                                  FROM ir_model_data
-                                 WHERE model = 'ir.model.fields'
-                                   AND res_id = %s
-                                 GROUP BY module)
+                   WHERE model = 'ir.model.fields'
+                     AND res_id = %s
+                     AND module != '__export__'
+               RETURNING module
             """,
             [fid],
         )
-        try:
-            with savepoint(cr):
-                cr.execute("UPDATE ir_model_data SET name=%s WHERE model='ir.model.fields' AND res_id=%s", [name, fid])
-        except psycopg2.IntegrityError:
-            # duplicate key. May happen du to conflict between
-            # some_model.sub_id and some_model_sub.id
-            # (before saas~11.2, where pattern changed)
-            name = "%s_%s" % (name, fid)
-            cr.execute("UPDATE ir_model_data SET name=%s WHERE model='ir.model.fields' AND res_id=%s", [name, fid])
+        if cr.rowcount:
+            modules = {mod for (mod,) in cr.fetchall()}
+
+            inserted = psycopg2.extras.execute_values(
+                cr,
+                """
+                INSERT INTO ir_model_data(module, name, model, res_id, noupdate)
+                     VALUES %s
+                ON CONFLICT DO NOTHING
+                  RETURNING module
+                """,
+                [(module, name, 'ir.model.fields', fid, True) for module in modules],
+                fetch=True,
+            )
+            conflicted = modules - {mod for (mod,) in inserted}
+            if conflicted:
+                # duplicate key. May happen du to conflict between
+                # some_model.sub_id and some_model_sub.id
+                # (before saas~11.2, where pattern changed)
+                on_conflict_name = "{}_{}".format(name, fid)
+                psycopg2.extras.execute_values(
+                    cr,
+                    """
+                    INSERT INTO ir_model_data(module, name, model, res_id, noupdate)
```

Comment:
Why do we insert conflicted modules? Shouldn't we update instead as we did before?

Reply:
I now remove all the existing xmlids. Before we were keeping one that we updated.

Reply:
Indeed. Got it. Looks good then.